### ISSUE TITLE
Easily enable valgrind in tests

### ISSUE THREAD

**Issue by richsalz** [openssl/openssl#8864]
One of the most useful features about the "memdebug" stuff is that it is easy to run the tests suite and find memory errors.  If we could easily run valgrind ```make test xxx=valgrind``` or a configure flag or such, it would reduce the need to build with that not-loved internal facility. :)  The hard part will be filtering the output to remove noise I think.

**Comment by tmshort** [comment]
Valgrind platform support: http://valgrind.org/info/platforms.html

Linux, Darwin, Android and Solaris (on Intel) are first-class citizens. It would be great to get this into the CI, but we'd have to be choosy.

**Comment by mattcaswell** [comment]
> It would be great to get this into the CI, but we'd have to be choosy.

Well we already have it in the CIs (well the equivalent of it) with "enable-crypto-mdebug"...but if you take that option away then we will need this.

**Comment by richsalz** [comment]
This issue doesn't say take away mdebug.  This issue is about making it easy to run all tests with valgrind :)

**Comment by tmshort** [comment]
In terms of "choosy", I mean that VMS (@levitte's platform), Windows and some of the BSDs are unsupported.

**Comment by levitte** [comment]
Note that this can be done on the (posix) command line already:

```
: ; make test EXE_SHELL="$(pwd)/util/shlib_wrap.sh valgrind"
```

**Comment by richsalz** [comment]
That would be cool, but:
```
DEV ~/openssl/github$ echo $pwd
/home/rsalz/openssl/github
DEV ~/openssl/github$ /bin/pwd
/home/rsalz/openssl/github
DEV ~/openssl/github$ make test EXE_SHELL="$pwd/util/shlib_wrap.sh valgrind" | head -30
make depend && make _tests
make[1]: Entering directory '/home/rsalz/openssl/github'
make[1]: Leaving directory '/home/rsalz/openssl/github'
make[1]: Entering directory '/home/rsalz/openssl/github'
( cd test; \
  mkdir -p test-runs; \
  SRCTOP=../. \
  BLDTOP=../. \
  RESULT_D=test-runs \
  PERL="/usr/bin/perl" \
  EXE_EXT= \
  OPENSSL_ENGINES=`cd .././engines 2>/dev/null && pwd` \
  OPENSSL_DEBUG_MEMORY=on \
    /usr/bin/perl .././test/run_tests.pl  )
../test/recipes/01-test_abort.t .................... ok
../test/recipes/01-test_sanity.t ................... 
Dubious, test returned 1 (wstat 256, 0x100)
Failed 1/1 subtests 
../test/recipes/01-test_symbol_presence.t .......... ok
../test/recipes/01-test_test.t ..................... 
Dubious, test returned 1 (wstat 256, 0x100)
Failed 1/1 subtests 
../test/recipes/02-test_errstr.t ................... 
Dubious, test returned 76 (wstat 19456, 0x4c00)
Failed 76/76 subtests 
../test/recipes/02-test_internal_context.t ......... 
Dubious, test returned 1 (wstat 256, 0x100)
Failed 1/1 subtests 
../test/recipes/02-test_internal_ctype.t ........... 
Dubious, test returned 1 (wstat 256, 0x100)
...etc...
```

**Comment by tmshort** [comment]
It worked for me without errors (but it's _SLOW_). Looks like you may have missed the parens around pwd... `EXE_SHELL="$(pwd)/util/shlib_wrap.sh valgrind"`, it seems they make a difference.

**Comment by tmshort** [comment]
I'm getting a few errors now (like I said... it's _SLOW_)...

I had memcheck cause a few system errors (likely bad instruction?).
We might need to specify `OPENSSL_ia32cap` to be compatible with Valgrind.
But what's also needed is a way to discover the errors, since they are reported in a way the test framework doesn't recognize.

**Comment by levitte** [comment]
> Looks like you may have missed the parens around pwd... `EXE_SHELL="$(pwd)/util/shlib_wrap.sh valgrind"`, it seems they make a difference.

Yes they do!  Sorry, that was a bashism, with the more posixly syntax, that would be:

```
: ; make test EXE_SHELL="`pwd`/util/shlib_wrap.sh valgrind"
```

**Comment by levitte** [comment]
To get errors for the leaks, you need to add ` --error-exitcode=1` (any exit code but 0, really) to that EXE_SHELL string.

**Comment by richsalz** [comment]
still fails for me, perhaps it's an environment problem.

**Comment by levitte** [comment]
Can you very simply do this, just to show us what happens?

```
make test EXE_SHELL="`pwd`/util/shlib_wrap.sh valgrind" TESTS=test_sanity V=1
```

**Comment by levitte** [comment]
(and of course paste the full resulting output, I hope that goes without saying)

**Comment by levitte** [comment]
Re slowness, I think there are ways to select what valgrind modules are used.  I think it's *all* by default...

**Comment by tmshort** [comment]
Default is just _memcheck_ per http://valgrind.org/docs/manual/quick-start.html#quick-start.mcrun.

**Comment by richsalz** [comment]
I had an alias for "pwd" :(

BTW, these flags seem useful ```--leak-check=full -q --exit-code=1```

**Comment by richsalz** [comment]
Perhaps this can be closed once #8867 is merged.
